ONVIF backchannel: Require-header gating, bare-track-URL fallback

Rebased onto live555 2026.07.23 with the series trimmed to runtime
patches only: the whitespace-normalization hunks (which depended on
the dropped warning-fix patch) are gone, functional content is
unchanged from the previous revision.

diff -ruN a/liveMedia/include/ServerMediaSession.hh b/liveMedia/include/ServerMediaSession.hh
--- a/liveMedia/include/ServerMediaSession.hh	2026-07-29 15:48:59.513539708 -0700
+++ b/liveMedia/include/ServerMediaSession.hh	2026-07-29 15:48:59.520244761 -0700
@@ -44,6 +44,7 @@
                               ServerMediaSession*& resultSession);
 
   char* generateSDPDescription(int addressFamily); // based on the entire session
+  char* generateSDPDescription(int addressFamily, char const* requireHeader); // conditional on Require:
       // Note: The caller is responsible for freeing the returned string
 
   char const* streamName() const { return fStreamName; }
@@ -186,6 +187,15 @@
   char const* rangeSDPLine() const;
       // returns a string to be delete[]d
 
+public:
+  // ONVIF backchannel support: subsessions can declare a "Require" tag.
+  // When set, the subsession's SDP lines are only included if the DESCRIBE
+  // request contains a matching "Require:" header value.
+  void setRequireTag(char const* tag);
+  char const* requireTag() const { return fRequireTag; }
+
+protected:
+
   ServerMediaSession* fParentSession;
   u_int32_t fSRTP_ROC; // horrible hack for SRTP; when the ROC changes, regenerate the SDP
 
@@ -196,6 +206,7 @@
 
   unsigned fTrackNumber; // within an enclosing ServerMediaSession
   char const* fTrackId;
+  char* fRequireTag; // if non-NULL, subsession requires this tag in Require: header
 };
 
 #endif
diff -ruN a/liveMedia/RTSPServer.cpp b/liveMedia/RTSPServer.cpp
--- a/liveMedia/RTSPServer.cpp	2026-07-29 15:48:59.514611537 -0700
+++ b/liveMedia/RTSPServer.cpp	2026-07-29 15:48:59.519688366 -0700
@@ -24,6 +24,9 @@
 #include "Base64.hh"
 #include <GroupsockHelper.hh>
 
+// Forward declaration (defined later in this file)
+static void lookForHeader(char const* headerName, char const* source, unsigned sourceLen, char* resultStr, unsigned resultMaxSize);
+
 ////////// RTSPServer implementation //////////
 
 RTSPServer*
@@ -33,7 +36,7 @@
   int ourSocketIPv4 = setUpOurSocket(env, ourPort, AF_INET);
   int ourSocketIPv6 = setUpOurSocket(env, ourPort, AF_INET6);
   if (ourSocketIPv4 < 0 && ourSocketIPv6 < 0) return NULL;
-  
+
   return new RTSPServer(env, ourSocketIPv4, ourSocketIPv6, ourPort, authDatabase, reclamationSeconds);
 }
 
@@ -41,15 +44,15 @@
 				 char const* name,
 				 RTSPServer*& resultServer) {
   resultServer = NULL; // unless we succeed
-  
+
   Medium* medium;
   if (!Medium::lookupByName(env, name, medium)) return False;
-  
+
   if (!medium->isRTSPServer()) {
     env.setResultMsg(name, " is not a RTSP server");
     return False;
   }
-  
+
   resultServer = (RTSPServer*)medium;
   return True;
 }
@@ -59,10 +62,10 @@
 	  int clientSocket, Boolean useIPv6) const {
   char* urlPrefix = rtspURLPrefix(clientSocket, useIPv6);
   char const* sessionName = serverMediaSession->streamName();
-  
+
   char* resultURL = new char[strlen(urlPrefix) + strlen(sessionName) + 1];
   sprintf(resultURL, "%s%s", urlPrefix, sessionName);
-  
+
   delete[] urlPrefix;
   return resultURL;
 }
@@ -85,9 +88,9 @@
 
     getsockname(clientSocket, (struct sockaddr*)&ourAddress, &namelen);
   }
-  
+
   char urlBuffer[100]; // more than big enough for "rtsp://<ip-address>:<port>/"
-  
+
   char const* addressPrefixInURL = ourAddress.ss_family == AF_INET6 ? "[" : "";
   char const* addressSuffixInURL = ourAddress.ss_family == AF_INET6 ? "]" : "";
 
@@ -102,14 +105,14 @@
 	    fOurConnectionsUseTLS ? "s" : "",
 	    addressPrefixInURL, AddressString(ourAddress).val(), addressSuffixInURL, portNumHostOrder);
   }
-  
+
   return strDup(urlBuffer);
 }
 
 UserAuthenticationDatabase* RTSPServer::setAuthenticationDatabase(UserAuthenticationDatabase* newDB) {
   UserAuthenticationDatabase* oldDB = fAuthDB;
   fAuthDB = newDB;
-  
+
   return oldDB;
 }
 
@@ -128,7 +131,7 @@
 							 incomingConnectionHandlerHTTPIPv6, this);
     return True;
   }
-  
+
   return False;
 }
 
@@ -208,17 +211,17 @@
   ::closeSocket(fHTTPServerSocketIPv4);
   envir().taskScheduler().turnOffBackgroundReadHandling(fHTTPServerSocketIPv6);
   ::closeSocket(fHTTPServerSocketIPv6);
-  
+
   cleanup(); // Removes all "ClientSession" and "ClientConnection" objects, and their tables.
   delete fClientConnectionsForHTTPTunneling;
-  
+
   // Delete any pending REGISTER requests:
   RTSPRegisterOrDeregisterSender* r;
   while ((r = (RTSPRegisterOrDeregisterSender*)fPendingRegisterOrDeregisterRequests->getFirst()) != NULL) {
     delete r;
   }
   delete fPendingRegisterOrDeregisterRequests;
-  
+
   // Empty out and close "fTCPStreamingDatabase":
   streamingOverTCPRecord* sotcp;
   while ((sotcp = (streamingOverTCPRecord*)fTCPStreamingDatabase->getFirst()) != NULL) {
@@ -279,7 +282,7 @@
     sotcp = sotcp->fNext;
   } while (sotcp != NULL);
   if (sotcp == NULL) return; // not found
-  
+
   if (sotcp == sotcpHead) {
     // We found it at the head of the list.  Remove it and reinsert the tail into the hash table:
     sotcpHead = sotcp->fNext;
@@ -344,7 +347,7 @@
     fOurRTSPServer.fClientConnectionsForHTTPTunneling->Remove(fOurSessionCookie);
     delete[] fOurSessionCookie;
   }
-  
+
   closeSocketsRTSP();
   delete[] fCurrentCSeq;
 }
@@ -403,12 +406,12 @@
     strcat(urlTotalSuffix, "/");
   }
   strcat(urlTotalSuffix, urlSuffix);
-    
+
   if (!authenticationOK("DESCRIBE", urlTotalSuffix, fullRequestStr)) return;
-    
+
   // We should really check that the request contains an "Accept:" #####
   // for "application/sdp", because that's what we're sending back #####
-    
+
   // Begin by looking up the "ServerMediaSession" object for the specified "urlTotalSuffix":
   ServerConnectionPair* scPair = new ServerConnectionPair(fOurRTSPServer, id());
   fOurServer.lookupServerMediaSession(urlTotalSuffix, DESCRIBELookupCompletionFunction, scPair);
@@ -437,13 +440,23 @@
       handleCmd_notFound();
       break;
     }
-    
+
     // Increment the "ServerMediaSession" object's reference count, in case someone removes it
     // while we're using it:
     session->incrementReferenceCount();
 
+    // Parse the "Require:" header from the original request (if any).
+    // This is used for ONVIF backchannel support: subsessions tagged with a
+    // require tag are only included in the SDP when the client sends the
+    // matching Require: header.
+    char requireHeader[256];
+    requireHeader[0] = '\0';
+    lookForHeader("Require", (char const*)fRequestBuffer, fRequestBytesAlreadySeen,
+		  requireHeader, sizeof requireHeader);
+
     // Then, assemble a SDP description for this session:
-    sdpDescription = session->generateSDPDescription(fAddressFamily);
+    sdpDescription = session->generateSDPDescription(fAddressFamily,
+						       requireHeader[0] != '\0' ? requireHeader : NULL);
     if (sdpDescription == NULL) {
       // This usually means that a file name that was specified for a
       // "ServerMediaSubsession" does not exist.
@@ -451,11 +464,11 @@
       break;
     }
     unsigned sdpDescriptionSize = strlen(sdpDescription);
-    
+
     // Also, generate our RTSP URL, for the "Content-Base:" header
     // (which is necessary to ensure that the correct URL gets used in subsequent "SETUP" requests).
     rtspURL = fOurRTSPServer.rtspURL(session, fClientInputSocket);
-    
+
     snprintf((char*)fResponseBuffer, sizeof fResponseBuffer,
 	     "RTSP/1.0 200 OK\r\nCSeq: %s\r\n"
 	     "%s"
@@ -469,7 +482,7 @@
 	     sdpDescriptionSize,
 	     sdpDescription);
   } while (0);
-  
+
   if (session != NULL) {
     // Decrement its reference count, now that we're done using it:
     session->decrementReferenceCount();
@@ -550,7 +563,7 @@
   // This parser is currently rather dumb; it should be made smarter #####
   char const* reqStr = (char const*)fRequestBuffer;
   unsigned const reqStrSize = fRequestBytesAlreadySeen;
-  
+
   // Read everything up to the first space as the command name:
   Boolean parseSucceeded = False;
   unsigned i;
@@ -560,12 +573,12 @@
       parseSucceeded = True;
       break;
     }
-    
+
     resultCmdName[i] = c;
   }
   resultCmdName[i] = '\0';
   if (!parseSucceeded) return False;
-  
+
   // Look for the string "HTTP/", before the first \r or \n:
   parseSucceeded = False;
   for (; i < reqStrSize-5 && reqStr[i] != '\r' && reqStr[i] != '\n'; ++i) {
@@ -576,22 +589,22 @@
     }
   }
   if (!parseSucceeded) return False;
-  
+
   // Get the 'URL suffix' that occurred before this:
   unsigned k = i-6;
   while (k > 0 && reqStr[k] == ' ') --k; // back up over white space
   unsigned j = k;
   while (j > 0 && reqStr[j] != ' ' && reqStr[j] != '/') --j;
   // The URL suffix is in position (j,k]:
-  if (k - j + 1 > urlSuffixMaxSize) return False; // there's no room> 
+  if (k - j + 1 > urlSuffixMaxSize) return False; // there's no room>
   unsigned n = 0;
   while (++j <= k) urlSuffix[n++] = reqStr[j];
   urlSuffix[n] = '\0';
-  
+
   // Look for various headers that we're interested in:
   lookForHeader("x-sessioncookie", &reqStr[i], reqStrSize-i, sessionCookie, sessionCookieMaxSize);
   lookForHeader("Accept", &reqStr[i], reqStrSize-i, acceptStr, acceptStrMaxSize);
-  
+
   return True;
 }
 
@@ -637,7 +650,7 @@
 #ifdef DEBUG
   fprintf(stderr, "Handled HTTP \"GET\" request (client output socket: %d)\n", fClientOutputSocket);
 #endif
-  
+
   // Construct our response:
   snprintf((char*)fResponseBuffer, sizeof fResponseBuffer,
 	   "HTTP/1.0 200 OK\r\n"
@@ -667,7 +680,7 @@
 #ifdef DEBUG
   fprintf(stderr, "Handled HTTP \"POST\" request (client input socket: %d)\n", fClientInputSocket);
 #endif
-  
+
   // Change the previous "RTSPClientSession" object's input socket to ours.  It will be used for subsequent requests:
   prevClientConnection->changeClientInputSocket(fClientInputSocket, fInputTLS,
 						extraData, extraDataSize);
@@ -699,7 +712,7 @@
     ::closeSocket(fClientOutputSocket);
   }
   fClientOutputSocket = -1;
-  
+
   closeSockets(); // closes fClientInputSocket
 }
 
@@ -727,7 +740,7 @@
 void RTSPServer::RTSPClientConnection::handleRequestBytes(int newBytesRead) {
   int numBytesRemaining = 0;
   ++fRecursionCount;
-  
+
   do {
     RTSPServer::RTSPClientSession* clientSession = NULL;
 
@@ -740,7 +753,7 @@
       fIsActive = False;
       break;
     }
-    
+
     Boolean endOfMsg = False;
     unsigned char* ptr = &fRequestBuffer[fRequestBytesAlreadySeen];
 #ifdef DEBUG
@@ -748,11 +761,11 @@
     fprintf(stderr, "RTSPClientConnection[%p]::handleRequestBytes() %s %d new bytes:%s\n",
 	    this, numBytesRemaining > 0 ? "processing" : "read", newBytesRead, ptr);
 #endif
-    
+
     if (fClientOutputSocket != fClientInputSocket && numBytesRemaining == 0) {
       // We're doing RTSP-over-HTTP tunneling, and input commands are assumed to have been Base64-encoded.
       // We therefore Base64-decode as much of this new data as we can (i.e., up to a multiple of 4 bytes).
-      
+
       // But first, we remove any whitespace that may be in the input data:
       unsigned toIndex = 0;
       for (int fromIndex = 0; fromIndex < newBytesRead; ++fromIndex) {
@@ -762,7 +775,7 @@
 	}
       }
       newBytesRead = toIndex;
-      
+
       unsigned numBytesToDecode = fBase64RemainderCount + newBytesRead;
       unsigned newBase64RemainderCount = numBytesToDecode%4;
       numBytesToDecode -= newBase64RemainderCount;
@@ -775,21 +788,21 @@
 	for (unsigned k = 0; k < decodedSize; ++k) fprintf(stderr, "%c", decodedBytes[k]);
 	fprintf(stderr, "\n");
 #endif
-	
+
 	// Copy the new decoded bytes in place of the old ones (we can do this because there are fewer decoded bytes than original):
 	unsigned char* to = ptr-fBase64RemainderCount;
 	for (unsigned i = 0; i < decodedSize; ++i) *to++ = decodedBytes[i];
-	
+
 	// Then copy any remaining (undecoded) bytes to the end:
 	for (unsigned j = 0; j < newBase64RemainderCount; ++j) *to++ = (ptr-fBase64RemainderCount+numBytesToDecode)[j];
-	
+
 	newBytesRead = decodedSize - fBase64RemainderCount + newBase64RemainderCount;
 	  // adjust to allow for the size of the new decoded data (+ remainder)
 	delete[] decodedBytes;
       }
       fBase64RemainderCount = newBase64RemainderCount;
     }
-    
+
     unsigned char* tmpPtr = fLastCRLF + 2;
     if (fBase64RemainderCount == 0) { // no more Base-64 bytes remain to be read/decoded
       // Look for the end of the message: <CR><LF><CR><LF>
@@ -805,12 +818,12 @@
 	++tmpPtr;
       }
     }
-    
+
     fRequestBufferBytesLeft -= newBytesRead;
     fRequestBytesAlreadySeen += newBytesRead;
-    
+
     if (!endOfMsg) break; // subsequent reads will be needed to complete the request
-    
+
     // Parse the request string into command name and 'CSeq', then handle the command:
     fRequestBuffer[fRequestBytesAlreadySeen] = '\0';
     char cmdName[RTSP_PARAM_STRING_MAX];
@@ -843,8 +856,8 @@
       fprintf(stderr, "parseRTSPRequestString() succeeded, returning cmdName \"%s\", urlPreSuffix \"%s\", urlSuffix \"%s\", CSeq \"%s\", Content-Length %u, with %d bytes following the message.\n", cmdName, urlPreSuffix, urlSuffix, cseq, contentLength, ptr + newBytesRead - (tmpPtr + 2));
 #endif
       // If there was a "Content-Length:" header, then make sure we've received all of the data that it specified:
-      if (ptr + newBytesRead < tmpPtr + 2 + contentLength) break; // we still need more data; subsequent reads will give it to us 
-      
+      if (ptr + newBytesRead < tmpPtr + 2 + contentLength) break; // we still need more data; subsequent reads will give it to us
+
       // If the request included a "Session:" id, and it refers to a client session that's
       // current ongoing, then use this command to indicate 'liveness' on that client session:
       Boolean const requestIncludedSessionId = sessionIdStr[0] != '\0';
@@ -853,7 +866,7 @@
 	  = (RTSPServer::RTSPClientSession*)(fOurRTSPServer.lookupClientSession(sessionIdStr));
 	if (clientSession != NULL) clientSession->noteLiveness();
       }
-    
+
       // We now have a complete RTSP request.
       // Handle the specified command (beginning with commands that are session-independent):
       delete[] fCurrentCSeq; fCurrentCSeq = strDup(cseq);
@@ -1004,7 +1017,7 @@
 	handleCmd_bad();
       }
     }
-    
+
 #ifdef DEBUG
     fprintf(stderr, "sending response: %s", fResponseBuffer);
 #endif
@@ -1014,25 +1027,25 @@
     } else {
         send(fClientOutputSocket, (char const*)fResponseBuffer, numBytesToWrite, MSG_NOSIGNAL);
    }
-    
+
     if (playAfterSetup) {
       // The client has asked for streaming to commence now, rather than after a
       // subsequent "PLAY" command.  So, simulate the effect of a "PLAY" command:
       clientSession->handleCmd_withinSession(this, "PLAY", urlPreSuffix, urlSuffix, (char const*)fRequestBuffer);
     }
-    
+
     // Check whether there are extra bytes remaining in the buffer, after the end of the request (a rare case).
     // If so, move them to the front of our buffer, and keep processing it, because it might be a following, pipelined request.
     unsigned requestSize = (fLastCRLF+4-fRequestBuffer) + contentLength;
     numBytesRemaining = fRequestBytesAlreadySeen - requestSize;
     resetRequestBuffer(); // to prepare for any subsequent request
-    
+
     if (numBytesRemaining > 0) {
       memmove(fRequestBuffer, &fRequestBuffer[requestSize], numBytesRemaining);
       newBytesRead = numBytesRemaining;
     }
   } while (numBytesRemaining > 0);
-  
+
   --fRecursionCount;
   // If it has a scheduledDelayedTask, don't delete the instance or close the sockets. The sockets can be reused in the task.
   if (!fIsActive && fScheduledDelayedTask <= 0) {
@@ -1052,14 +1065,14 @@
 					char const*& response) {
   // Initialize the result parameters to default values:
   username = realm = nonce = uri = response = NULL;
-  
+
   // First, find "Authorization:"
   while (1) {
     if (*buf == '\0') return False; // not found
     if (_strncasecmp(buf, "Authorization: Digest ", 22) == 0) break;
     ++buf;
   }
-  
+
   // Then, run through each of the fields, looking for ones we handle:
   char const* fields = buf + 22;
   char* parameter = strDupSize(fields);
@@ -1108,16 +1121,16 @@
   // If we weren't set up with an authentication database, we're OK:
   UserAuthenticationDatabase* authDB = fOurRTSPServer.getAuthenticationDatabaseForCommand(cmdName);
   if (authDB == NULL) return True;
-  
+
   char const* username = NULL; char const* realm = NULL; char const* nonce = NULL;
   char const* uri = NULL; char const* response = NULL;
   Boolean success = False;
-  
+
   do {
     // To authenticate, we first need to have a nonce set up
     // from a previous attempt:
     if (fCurrentAuthenticator.nonce() == NULL) break;
-    
+
     // Next, the request needs to contain an "Authorization:" header,
     // containing a username, (our) realm, (our) nonce, uri,
     // and response string:
@@ -1129,7 +1142,7 @@
 	|| uri == NULL || response == NULL) {
       break;
     }
-    
+
     // Next, the username has to be known to us:
     char const* password = authDB->lookupPassword(username);
 #ifdef DEBUG
@@ -1137,7 +1150,7 @@
 #endif
     if (password == NULL) break;
     fCurrentAuthenticator.setUsernameAndPassword(username, password, authDB->passwordsAreMD5());
-    
+
     // Finally, compute a digest response from the information that we have,
     // and compare it to the one that we were given:
     char const* ourResponse
@@ -1145,10 +1158,10 @@
     success = (strcmp(ourResponse, response) == 0);
     fCurrentAuthenticator.reclaimDigestResponse(ourResponse);
   } while (0);
-  
+
   delete[] (char*)realm; delete[] (char*)nonce;
   delete[] (char*)uri; delete[] (char*)response;
-  
+
   if (success) {
     // The user has been authenticated.
     // Now allow subclasses a chance to validate the user against the IP address and/or URL suffix.
@@ -1162,7 +1175,7 @@
   }
   delete[] (char*)username;
   if (success) return True;
-  
+
   // If we get here, we failed to authenticate the user.
   // Send back a "401 Unauthorized" response, with a new random nonce:
   Boolean isInitial401 = fCurrentAuthenticator.nonce() == NULL;
@@ -1209,7 +1222,7 @@
 ::setRTSPResponse(char const* responseStr, char const* contentStr) {
   if (contentStr == NULL) contentStr = "";
   unsigned const contentLen = strlen(contentStr);
-  
+
   snprintf((char*)fResponseBuffer, sizeof fResponseBuffer,
 	   "RTSP/1.0 %s\r\n"
 	   "CSeq: %s\r\n"
@@ -1227,7 +1240,7 @@
 ::setRTSPResponse(char const* responseStr, u_int32_t sessionId, char const* contentStr) {
   if (contentStr == NULL) contentStr = "";
   unsigned const contentLen = strlen(contentStr);
-  
+
   snprintf((char*)fResponseBuffer, sizeof fResponseBuffer,
 	   "RTSP/1.0 %s\r\n"
 	   "CSeq: %s\r\n"
@@ -1251,11 +1264,11 @@
   fClientInputSocket = newSocketNum;
   envir().taskScheduler().setBackgroundHandling(fClientInputSocket, SOCKET_READABLE|SOCKET_EXCEPTION,
 						incomingRequestHandler, this);
-  
+
   // Change the TLS state:
   fPOSTSocketTLS.assignStateFrom(*newTLSState);
   fInputTLS = &fPOSTSocketTLS;
-  
+
   // Also write any extra data to our buffer, and handle it:
   if (extraDataSize > 0 && extraDataSize <= fRequestBufferBytesLeft/*sanity check; should always be true*/) {
     unsigned char* ptr = &fRequestBuffer[fRequestBytesAlreadySeen];
@@ -1288,7 +1301,7 @@
     fStreamStates[trackNum].subsession->deleteStream(fOurSessionId, fStreamStates[trackNum].streamToken);
     fStreamStates[trackNum].subsession = NULL;
   }
-  
+
   // Optimization: If all subsessions have now been deleted, then we can delete ourself now:
   Boolean noSubsessionsRemain = True;
   for (unsigned i = 0; i < fNumStreamStates; ++i) {
@@ -1346,7 +1359,7 @@
     if (_strncasecmp(buf, "Transport:", 10) == 0) break;
     ++buf;
   }
-  
+
   // Then, run through each of the fields, looking for ones we handle:
   char const* fields = buf + 10;
   while (*fields == ' ') ++fields;
@@ -1422,7 +1435,7 @@
     if (_strncasecmp(buf, "x-playNow:", 10) == 0) break;
     ++buf;
   }
-  
+
   return True;
 }
 
@@ -1543,6 +1556,27 @@
   do {
     if (sms == NULL) {
       if (fOurServerMediaSession == NULL) {
+	// Fallback: Some ONVIF clients resolve relative track control URLs
+	// against the server root instead of the Content-Base, sending e.g.
+	// "SETUP rtsp://host/track4" instead of "rtsp://host/ch0/track4".
+	// When urlPreSuffix was empty and urlSuffix looks like a track id,
+	// search all sessions for a subsession with a matching trackId.
+	if (fURLPreSuffix[0] == '\0' && fURLSuffix[0] != '\0' && fTrackId == NULL) {
+	  ServerMediaSessionIterator smIter(fOurRTSPServer);
+	  ServerMediaSession* session;
+	  while ((session = smIter.next()) != NULL) {
+	    ServerMediaSubsessionIterator subIter(*session);
+	    ServerMediaSubsession* sub;
+	    while ((sub = subIter.next()) != NULL) {
+	      if (sub->trackId() != NULL && strcmp(fURLSuffix, sub->trackId()) == 0) {
+		// Found the track in this session — retry with it
+		fTrackId = fURLSuffix;
+		handleCmd_SETUP_afterLookup2(ourClientConnection, session);
+		return;
+	      }
+	    }
+	  }
+	}
 	// The client asked for a stream that doesn't exist (and this session descriptor has not been used before):
 	ourClientConnection->handleCmd_notFound();
       } else {
@@ -1561,12 +1595,12 @@
 	break;
       }
     }
-    
+
     if (fStreamStates == NULL) {
       // This is the first "SETUP" for this session.  Set up our array of states for all of this session's subsessions (tracks):
       fNumStreamStates = fOurServerMediaSession->numSubsessions();
       fStreamStates = new struct streamState[fNumStreamStates];
-      
+
       ServerMediaSubsessionIterator iter(*fOurServerMediaSession);
       ServerMediaSubsession* subsession;
       for (unsigned i = 0; i < fNumStreamStates; ++i) {
@@ -1587,7 +1621,7 @@
        }
 #endif
     }
-    
+
     // Look up information for the specified subsession (track):
     ServerMediaSubsession* subsession = NULL;
     unsigned trackNum;
@@ -1705,7 +1739,7 @@
 				    destinationAddress, destinationTTL, fIsMulticast,
 				    serverRTPPort, serverRTCPPort,
 				    fStreamStates[trackNum].streamToken);
-    
+
     AddressString destAddrStr(destinationAddress);
     AddressString sourceAddrStr(sourceAddr);
     char timeoutParameterString[100];
@@ -1822,7 +1856,7 @@
   //   or "urlPreSuffix" and "urlSuffix" are both nonempty, but when concatenated, (with "/") form the session (stream) name.
   // Begin by figuring out which of these it is:
   ServerMediaSubsession* subsession;
-  
+
   if (fOurServerMediaSession == NULL) { // There wasn't a previous SETUP!
     ourClientConnection->handleCmd_notSupported();
     return;
@@ -1852,11 +1886,27 @@
       ourClientConnection->handleCmd_notFound();
       return;
     }
+  } else if (urlPreSuffix[0] == '\0' && urlSuffix[0] != '\0') {
+    // Fallback: client sent a bare track URL (e.g. PLAY rtsp://host/track4)
+    // without the stream prefix. Try urlSuffix as a track id in our session.
+    ServerMediaSubsessionIterator iter(*fOurServerMediaSession);
+    subsession = NULL;
+    ServerMediaSubsession* sub;
+    while ((sub = iter.next()) != NULL) {
+      if (sub->trackId() != NULL && strcmp(sub->trackId(), urlSuffix) == 0) {
+	subsession = sub;
+	break;
+      }
+    }
+    if (subsession == NULL) {
+      ourClientConnection->handleCmd_notFound();
+      return;
+    }
   } else { // the request doesn't match a known stream and/or track at all!
     ourClientConnection->handleCmd_notFound();
     return;
   }
-  
+
   if (strcmp(cmdName, "TEARDOWN") == 0) {
     handleCmd_TEARDOWN(ourClientConnection, subsession, fullRequestStr);
   } else if (strcmp(cmdName, "PLAY") == 0) {
@@ -1887,9 +1937,9 @@
       }
     }
   }
-  
+
   setRTSPResponse(ourClientConnection, "200 OK");
-  
+
   // Optimization: If all subsessions have now been torn down, then we know that we can reclaim our object now.
   // (Without this optimization, however, this object would still get reclaimed later, as a result of a 'liveness' timeout.)
   Boolean noSubsessionsRemain = True;
@@ -1931,7 +1981,7 @@
     sprintf(buf, "Scale: %f\r\n", scale);
   }
   scaleHeader = strDup(buf);
-  
+
   // Parse the client's "Range:" header, if any:
   float duration = 0.0;
   double rangeStart = 0.0, rangeEnd = 0.0;
@@ -1939,7 +1989,7 @@
   Boolean startTimeIsNow;
   Boolean sawRangeHeader
     = parseRangeHeader(fullRequestStr, rangeStart, rangeEnd, absStart, absEnd, startTimeIsNow);
-  
+
   if (sawRangeHeader && absStart == NULL/*not seeking by 'absolute' time*/) {
     // Use this information, plus the stream's duration (if known), to create our own "Range:" header, for the response:
     duration = subsession == NULL /*aggregate op*/
@@ -1949,7 +1999,7 @@
       // Use the largest of these durations in our header
       duration = -duration;
     }
-    
+
     // Make sure that "rangeStart" and "rangeEnd" (from the client's "Range:" header)
     // have sane values, before we send back our own "Range:" header in our response:
     if (rangeStart < 0.0) rangeStart = 0.0;
@@ -1964,7 +2014,7 @@
       rangeEnd = tmp;
     }
   }
-  
+
   // Create a "RTP-Info:" line.  It will get filled in from each subsession's state:
   char const* rtpInfoFmt =
     "%s" // "RTP-Info:", plus any preceding rtpInfo items
@@ -1976,7 +2026,7 @@
   unsigned rtpInfoFmtSize = strlen(rtpInfoFmt);
   char* rtpInfo = strDup("RTP-Info: ");
   unsigned i, numRTPInfoItems = 0;
-  
+
   // Do any required seeking/scaling on each subsession, before starting streaming.
   // (However, we don't do this if the "PLAY" request was for just a single subsession
   // of a multiple-subsession stream; for such streams, seeking/scaling can be done
@@ -1989,11 +2039,11 @@
 	}
 	if (absStart != NULL) {
 	  // Special case handling for seeking by 'absolute' time:
-	
+
 	  fStreamStates[i].subsession->seekStream(fOurSessionId, fStreamStates[i].streamToken, absStart, absEnd);
 	} else {
 	  // Seeking by relative (NPT) time:
-	  
+
 	  u_int64_t numBytes;
 	  if (!sawRangeHeader || startTimeIsNow) {
 	    // We're resuming streaming without seeking, so we just do a 'null' seek
@@ -2017,14 +2067,14 @@
       }
     }
   }
-  
+
   // Create the "Range:" header that we'll send back in our response.
   // (Note that we do this after seeking, in case the seeking operation changed the range start time.)
   char* rangeHeader;
   if (absStart != NULL) {
     // We're seeking by 'absolute' time:
     char* rangeHeaderBuf;
-    
+
     if (absEnd == NULL) {
       rangeHeaderBuf = new char[100 + strlen(absStart)]; // ample space
       sprintf(rangeHeaderBuf, "Range: clock=%s-\r\n", absStart);
@@ -2062,7 +2112,7 @@
     }
     rangeHeader = strDup(buf);
   }
-  
+
   // Now, start streaming:
   for (i = 0; i < fNumStreamStates; ++i) {
     if (subsession == NULL /* means: aggregated operation */
@@ -2103,7 +2153,7 @@
     rtpInfo[rtpInfoLen+1] = '\n';
     rtpInfo[rtpInfoLen+2] = '\0';
   }
-  
+
   // Fill in the response:
   snprintf((char*)ourClientConnection->fResponseBuffer, sizeof ourClientConnection->fResponseBuffer,
 	   "RTSP/1.0 200 OK\r\n"
@@ -2137,7 +2187,7 @@
       }
     }
   }
-  
+
   setRTSPResponse(ourClientConnection, "200 OK", fOurSessionId);
 }
 
diff -ruN a/liveMedia/ServerMediaSession.cpp b/liveMedia/ServerMediaSession.cpp
--- a/liveMedia/ServerMediaSession.cpp	2026-07-29 15:48:59.515736641 -0700
+++ b/liveMedia/ServerMediaSession.cpp	2026-07-29 15:48:59.520186317 -0700
@@ -209,6 +209,10 @@
 }
 
 char* ServerMediaSession::generateSDPDescription(int addressFamily) {
+  return generateSDPDescription(addressFamily, NULL);
+}
+
+char* ServerMediaSession::generateSDPDescription(int addressFamily, char const* requireHeader) {
   struct sockaddr_storage ourAddress;
   if (addressFamily == AF_INET) {
     ourAddress.ss_family = AF_INET;
@@ -219,7 +223,7 @@
       ((sockaddr_in6&)ourAddress).sin6_addr.s6_addr[i] = ourIPv6Address(envir())[i];
     }
   }
-  
+
   AddressString ipAddressStr(ourAddress);
   unsigned ipAddressStrSize = strlen(ipAddressStr.val());
 
@@ -251,6 +255,11 @@
     ServerMediaSubsession* subsession;
     for (subsession = fSubsessionsHead; subsession != NULL;
 	 subsession = subsession->fNext) {
+      // Skip subsessions that require a specific tag not present in the request
+      char const* reqTag = subsession->requireTag();
+      if (reqTag != NULL) {
+	if (requireHeader == NULL || strstr(requireHeader, reqTag) == NULL) continue;
+      }
       char const* sdpLines = subsession->sdpLines(addressFamily);
       if (sdpLines == NULL) continue; // the media's not available
       sdpLength += strlen(sdpLines);
@@ -316,6 +325,11 @@
     char* mediaSDP = sdp;
     for (subsession = fSubsessionsHead; subsession != NULL;
 	 subsession = subsession->fNext) {
+      // Skip subsessions that require a specific tag not present in the request
+      char const* reqTag = subsession->requireTag();
+      if (reqTag != NULL) {
+	if (requireHeader == NULL || strstr(requireHeader, reqTag) == NULL) continue;
+      }
       unsigned mediaSDPLength = strlen(mediaSDP);
       mediaSDP += mediaSDPLength;
       sdpLength -= mediaSDPLength;
@@ -359,14 +373,20 @@
 
 ServerMediaSubsession::ServerMediaSubsession(UsageEnvironment& env)
   : Medium(env),
-    fParentSession(NULL), fSRTP_ROC(0), fNext(NULL), fTrackNumber(0), fTrackId(NULL) {
+    fParentSession(NULL), fSRTP_ROC(0), fNext(NULL), fTrackNumber(0), fTrackId(NULL), fRequireTag(NULL) {
 }
 
 ServerMediaSubsession::~ServerMediaSubsession() {
   delete[] (char*)fTrackId;
+  delete[] fRequireTag;
   Medium::close(fNext);
 }
 
+void ServerMediaSubsession::setRequireTag(char const* tag) {
+  delete[] fRequireTag;
+  fRequireTag = tag ? strDup(tag) : NULL;
+}
+
 char const* ServerMediaSubsession::trackId() {
   if (fTrackNumber == 0) return NULL; // not yet in a ServerMediaSession
 
